packages/dds/tree/src/core/edit-manager/editManager.ts

```diff
@@ -113,119 +162,139 @@ export class EditManager<
-		return (this.getLastCommit() ?? fail("No sequenced changes")).changeset;
+		return (this.getLastCommit() ?? fail("No sequenced changes")).change;
 	}
 
 	public getLastCommit(): Commit<TChangeset> | undefined {
-		return this.trunk[this.trunk.length - 1];
+		return this.trunk;
 	}
 
 	public getLocalChanges(): readonly RecursiveReadonly<TChangeset>[] {
-		return this.localChanges.map((change) => change.change);
+		return getPathFromBase(this.localBranch, this.trunk).map((c) => c.change);
 	}
 
-	public addSequencedChange(newCommit: Commit<TChangeset>): Delta.Root {
+	public addSequencedChange(
+		newCommit: Commit<TChangeset>,
+		sequenceNumber: SeqNumber,
+		referenceSequenceNumber: SeqNumber,
+	): Delta.Root {
 		assert(
 			this.localSessionId !== undefined,
 			0x429 /* The session ID should be set before processing changes */,
 		);
 
-		if (this.trunk.length > 0) {
-			const lastSeqNumber = this.trunk[this.trunk.length - 1].seqNumber;
-			assert(
-				newCommit.seqNumber > lastSeqNumber,
-				0x3a2 /* Incoming remote op sequence# <= local collabWindow's currentSequence# */,
-			);
-		}
 		if (newCommit.sessionId === this.localSessionId) {
 			// `newCommit` should correspond to the oldest change in `localChanges`, so we move it into trunk.
 			// `localChanges` are already rebased to the trunk, so we can use the stored change instead of rebasing the
 			// change in the incoming commit.
-			const change =
-				this.localChanges.shift() ??
-				fail(
-					"Received a sequenced change from the local session despite having no local changes",
-				);
-
-			// TODO: The local change may contain references to local revision tags.
-			// When other clients rebase this change, they will instead use the corresponding sequence numbers
-			// instead of this client's local revision numbers. Although this need not inherently cause a divergence
-			// in the state of the document, this is potentially dangerous.
-			// This could be solved by one of the following approaches:
-			// A) Maintain a `Branch` of sequenced changes from this client in addition to `localChanges`,
-			//    and only use the rebased changes from that branch when adding to the trunk.
-			// B) Use a UUID as the revision tag for each changeset instead of a sequence number,
-			//    so that we do not need local revision tags.
-			// C) Strip any references to local changes before adding the change to the trunk.
-			//    This would also have to be done on each branch with references to changes local to that branch.
-			this.trunk.push({
-				...newCommit,
-				changeset: change.change,
-			});
+			const localPath = getPathFromBase(this.localBranch, this.trunk);
+			// Get the first revision in the local branch, and then remove it
+			const { change } = localPath.shift() ?? fail(UNEXPECTED_SEQUENCED_LOCAL_EDIT);
+			this.pushToTrunk(sequenceNumber, { ...newCommit, change });
+			this.localBranch = localPath.reduce(mintCommit, this.trunk);
```

Comment:
Is the fact that `GraphCommit.parent` reasonly the only reason we can't do something like this instead:
```
if (localPath.length > 0) {
	localPath[0].parent = this.trunk;
} else {
	this.localBranch = this.trunk;
}
```
And if so, do we really need this immutability in this case?
I might be nudging us towards premature optimization, but it does worry me a bit that in the common case of a single writer pushing a lot of edits, we'll end up perform (and allocating) O(n²) for n edits.

Reply:
I think you might be right. I'm currently planning on our async transaction branches being branches off of the local branch (i.e. "local local" branches :P) and at first glance it seemed to me that mutating the local branch via `localPath[0].parent = this.trunk` might cause some inconsistency but... maybe it doesn't. I think we need to figure out how "local local" branches work first (e.g. do they get eagerly rebased along with the local branch, or can they fall behind?). Is it alright if I put a todo here to attempt to optimize this in the next branching PR?

Reply:
Sounds good.